### PR TITLE
Derive Eq

### PR DESCRIPTION
**Pull request by mlutze**
Fixes https://github.com/flix/flix/issues/2157

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Deriver.scala [hunk 1]
```diff
@@ -167,6 +170,121 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
       KindedAst.MatchRule(pat, guard, exp)
   }
 
+  /**
+    * Creates a toString instance for the given enum.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Deriver.scala:174]: Eq instance.

File: main/src/ca/uwaterloo/flix/language/phase/Deriver.scala [hunk 2]
```diff
@@ -167,6 +170,121 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
...
+    *   pub def eq(x: E[a], y: E[a]): Bool = (x, y) match {
+    *     case (C0, C0) => true
+    *     case (C1(x0), C1(y0)) => x0 == y0
+    *     case (C2(x0, x1), C2(y0, y1)) => x0 == y0 and x1 == y1
+    *     case _ => false
+    *   }
+    * }
+    * }}}
+    */
+  private def createEq(enum: KindedAst.Enum, loc: SourceLocation, root: KindedAst.Root)(implicit flix: Flix): KindedAst.Instance = enum match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Deriver.scala:197]: Usually all of these functions are called `mkX`. (You even have an mk later! :))

File: main/src/ca/uwaterloo/flix/language/phase/Deriver.scala [hunk 3]
```diff
@@ -167,6 +170,121 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
...
+    case KindedAst.Enum(_, _, _, tparams, _, cases, _, sc, _) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Deriver.scala:198]: For both createEq and createToString could they not be subdivded to more closely follow the structure of the AST, e.g.:

- mkEqInstance calls
- mkEqSpec and mkEqDef

Makes sense?

File: main/src/ca/uwaterloo/flix/language/phase/Deriver.scala [hunk 4]
```diff
@@ -167,6 +170,121 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
...
+        mod = Ast.Modifiers.Empty,
+        sym = eqClassSym,
+        tpe = sc.base,
+        tconstrs = tconstrs,
+        defs = List(defn),
+        ns = Name.RootNS,
+        loc = loc
+      )
+  }
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Deriver.scala:246]: I find the code a bit hard to follow. It would be nice if the comments had some more of the fragments.

File: main/test/flix/Test.Derives.Eq.flix [hunk 5]
```diff
@@ -0,0 +1,115 @@
+namespace Test/Derives/Eq {
+    pub enum Enum with Eq {
```
Comment (magnus-madsen) [inline main/test/flix/Test.Derives.Eq.flix:2]: Tests look good, but I think we should also have a few tests that instantiate `PolyEnum` with a non-primitive type- e.g. one of the other enums.

(This could also apply to the toString stuff).

File: main/src/ca/uwaterloo/flix/language/phase/Deriver.scala (hunk 3 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Deriver.scala:198]: Yep, makes sense.

**Comment by magnus-madsen** [comment]
Still a draft?

**Comment by mlutze** [comment]
> Still a draft?

Yep

File: main/src/ca/uwaterloo/flix/language/phase/Deriver.scala (hunk 4 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Deriver.scala:250]: I added some more documentation especially to the match/case part. Let me know if more is needed.

**Comment by magnus-madsen** [comment]
Yay!
